data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbSourceCoordinationStore.java

```diff
@@ -23,33 +33,106 @@
+
     @Override
     public Optional<SourcePartitionStoreItem> getSourcePartitionItem(final String partitionKey) {
-        return Optional.empty();
+        return dynamoDbClientWrapper.getSourcePartitionItem(partitionKey);
     }
 
     @Override
     public boolean tryCreatePartitionItem(final String partitionKey,
                                           final SourcePartitionStatus sourcePartitionStatus,
                                           final Long closedCount,
                                           final String partitionProgressState) {
+        final DynamoDbSourcePartitionItem newPartitionItem = new DynamoDbSourcePartitionItem();
+        newPartitionItem.setSourcePartitionKey(partitionKey);
+        newPartitionItem.setSourcePartitionStatus(sourcePartitionStatus);
+        newPartitionItem.setClosedCount(closedCount);
+        newPartitionItem.setPartitionProgressState(partitionProgressState);
+        newPartitionItem.setVersion(0L);
+
+        final Optional<SourcePartitionStoreItem> optionalPartitionItem = dynamoDbClientWrapper.getSourcePartitionItem(partitionKey);
+        if (optionalPartitionItem.isEmpty()) {
+             return dynamoDbClientWrapper.tryCreatePartitionItem(newPartitionItem);
+        }
+
         return false;
     }
 
     @Override
-    public Optional<SourcePartitionStoreItem> tryAcquireAvailablePartition() {
+    public Optional<SourcePartitionStoreItem> tryAcquireAvailablePartition(final String ownerId, final Duration ownershipTimeout) {
+        final Optional<PageIterable<DynamoDbSourcePartitionItem>> dynamoDbSourcePartitionItemPageIterable =
+                dynamoDbClientWrapper.getSourcePartitionItems(Expression.builder()
+                        .expressionValues(Map.of(
+                                ":s", AttributeValue.builder().s(SourcePartitionStatus.UNASSIGNED.name()).build(),
+                                ":t", AttributeValue.builder().s(Instant.now().toString()).build(),
+                                ":ro", AttributeValue.builder().s(Instant.now().toString()).build(),
+                                ":null", AttributeValue.builder().nul(true).build()))
+                        .expression(AVAILABLE_PARTITIONS_FILTER_EXPRESSION)
+                        .build());
+
+        if (dynamoDbSourcePartitionItemPageIterable.isEmpty()) {
+            return Optional.empty();
+        }
+
+        try {
+            for (final DynamoDbSourcePartitionItem item : dynamoDbSourcePartitionItemPageIterable.get().items()) {
+                item.setPartitionOwner(ownerId);
+                item.setPartitionOwnershipTimeout(Instant.now().plus(ownershipTimeout));
+                item.setSourcePartitionStatus(SourcePartitionStatus.ASSIGNED);
+                final boolean acquired = tryAcquireItem(item);
+
+                if (acquired) {
+                    return Optional.of(item);
+                }
+            }
+        } catch (final ProvisionedThroughputExceededException e) {
```

Comment:
My understanding is that this will get thrown while iterating through the scan when we have to load more items than the limit of 20. I am not sure how this works in relation to the client exponential backoff and retry. Will it also backoff and retry loading the next page while iterating?